app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/LectureListTransformer.kt

```diff
@@ -0,0 +1,99 @@
+package nerd.tuxmobil.fahrplan.congress.repositories
+
+import android.support.v4.util.SparseArrayCompat
+import nerd.tuxmobil.fahrplan.congress.models.Lecture
+import nerd.tuxmobil.fahrplan.congress.models.RoomData
+import nerd.tuxmobil.fahrplan.congress.models.ScheduleData
+
+class LectureListTransformer(private val prioritizedRoomProvider: PrioritizedRoomProvider) {
+
+    /**
+     * Transform data loaded from the database to a data structure that we would like to use in
+     * UI code.
+     * Then transform it to the (legacy) data structure that the UI is currently using.
+     */
+    fun legacyTransformLectureList(dayIndex: Int, lectures: List<Lecture>): LegacyLectureData {
+        val scheduleData = transformLectureList(dayIndex, lectures)
+
+        // Set Lecture.roomIndex to match the index in ScheduleData.roomDataList
+        scheduleData.roomDataList.forEachIndexed { index, roomData ->
+            roomData.lectures.forEach { it.roomIndex = index }
+        }
+
+        // Map room title to room index
+        val roomsMap = scheduleData.roomDataList
+                .mapIndexed { index, roomData -> roomData.roomName to index }
+                .toMap()
+
+        // Map column index to room index
+        // This is simple identity map (0 -> 0, 1 -> 1, etc.)
+        // Room indices provided by the database are discarded here on purpose.
+        // Future UI code which assembles room columns should not be based on these room indices.
+        val roomList = SparseArrayCompat<Int>()
+        for (index in scheduleData.roomDataList.indices) {
+            roomList.put(index, index)
+        }
+
+        return LegacyLectureData(
+                lectureListDay = dayIndex,
+                lectureList = lectures,
```

Comment:
I ran the unit tests which I wrote for `FahrplanFragment#loadLectureList` against this new implementation. It showed that the resulting `lectureList` is not output sorted by the legacy method. See https://github.com/EventFahrplan/EventFahrplan/blob/4b80f7e1b046c421027784c681e9475250f20698/app/src/test/java/nerd/tuxmobil/fahrplan/congress/schedule/FahrplanFragmentLoadLectureListTest.kt#L251-L274

I assume `lectureList = lectures.sortedBy { it.dateUTC }.toList(),` is missing here.

Reply:
`Lecture.roomIndex` is rewritten by the new code.

Reply:
Yes. Here is the updated version of the test:

``` kotlin
@Test
fun `Sorts "lectureList" by "dateUTC" field`() {
    MyApp.lectureList = null
    val dayIndex = 0
    val lecture0 = DatabaseLecture(eventId = "L0", room = "Ada", roomIndex = 0, dateUTC = 200)
    val lecture1 = DatabaseLecture(eventId = "L1", room = "Ada", roomIndex = 0, dateUTC = 100)
    val lectures = listOf(lecture0, lecture1)
    whenever(lecturesDatabaseRepository.queryLecturesForDayIndexOrderedByDateUtc(dayIndex)) doReturn lectures

    FahrplanFragment.loadLectureList(NoLogging, testableAppRepository, dayIndex, false, onDoneDoNothing)

    val expectedLectures = listOf(
            Lecture("L1").apply {
                room = "Ada"
                roomIndex = -1 // Not part of Lecture#equals, therefore ignored.

                dateUTC = 100
            },
            Lecture("L0").apply {
                room = "Ada"
                roomIndex = -1 // Not part of Lecture#equals, therefore ignored.
                dateUTC = 200
            }
    )
    assertThat(MyApp.lectureList as List<*>).isEqualTo(expectedLectures)
}
```

Reply:
Right, sorting by date only happened in the new code. Added a test and updated `legacyTransformLectureList()` just now.